docs/reporting.md

```diff
@@ -187,4 +187,265 @@ There are basically two things to keep in mind when upgrading to `SoberSwag::Rep
+This is because *serialization* is the *opposite* of *parsing*.
+Kinda neat, huh?
+
+### Composite Types
+
+These types work with *one or more* inputs to build up *another*.
+
+- `SoberSwag::Reporting::Input::List`, which lets you parse a JSON array.
+  IE:
+  ```ruby
+  SoberSwag::Reporting::Input::List.of(SoberSwag::Reporting::Input.number)
+  ```
+  Lets you parse a list of numbers.
+- `SoberSwag::Reporting::Input::Either`, which lets you parse one input, and if that fails, parse another.
+  This represents a *choice* of input types.
+  This is best used via:
+  ```ruby
+  SoberSwag::Reporting::Input.text | SoberSwag::Reporting::Input.number
+  # or
+  SoberSwag::Reporting::Input.text.or SoberSwag::Reporting::Input.number
+  ```
+  This is useful if you want to allow multiple input formats.
+- `SoberSwag::Reporting::Input::Dictionary`, which lets you parse a JSON dictionary with arbitrary keys.
+  For example, to parse this JSON (assuming you don't know the keys ahead of time):
+  ```json
+  {
+    "mike": 100,
+    "bob": 1000,
+    "joey": 12,
+    "yes": 1213
+  }
+  ```
+  You can use:
+  ```ruby
+  SoberSwag::Reporting::Input::Dictionary.of(SoberSwag::Reporting::Input.number)
+  ```
+
+  This will parse to a Ruby hash, with string keys.
+  If you want symbols, you can simply use `.mapped`:
+  ```ruby
+  SoberSwag::Reporting::Input::Dictionary.of(
+    SoberSwag::Reporting::Input.number
+  ).mapped { |hash| hash.transform_keys(&:to_sym) }
+  ```
+  Pretty cool, right?
+- `SoberSwag::Reporting::Input::Enum`, which lets you parse an *enum value*.
+  This input will validate that the given value is in the enum.
+  Note that this doesn't only work with strings!
+  You can use:
+
+  ```ruby
+  SoberSwag::Reporting::Input.number.enum(-1, 0, 1)
+  ```
+
+  And things will work fine.
+
+### Validating Types
+
+These types provide *validation* on an input.
+The validations provided match the specifications in swagger.
+
+- `SoberSwag::Reporting::Input::InRange`, which specifies that a value should be *within a range*.
+  You can use it like:
+
+  ```ruby
+  SoberSwag::Reporting::Input::InRange.new(
+    SoberSwag::Reporting::Input::Number,
+    1..100
+  )
+  ```
+- `SoberSwag::Reporting::Input::MultipleOf`, which specifies that a number is a *multiple of* some other number.
+  You can use it like this:
+  ```ruby
+  SoberSwag::Reporting::Input.number.multiple_of(2)
+  ```
+
+  Note that the `#multiple_of` method is only available on the `SoberSwag::Reporting::Input::Number` class.
+- `SoberSwag::Reporting::Input::Pattern`, which lets you check that an input *matches a regexp*.
+  You can use it like:
+
+  ```ruby
+  SoberSwag::Reporting::Input.text.with_pattern(/\A(R|r)ich (E|e)vans\z/)
+  ```
+
+  Note that the `with_pattern` method is only available on `SoberSwag::Reporting::Input::Text`
+
+### Documentating Types
+
+These types allow you to add additional documentation.
```

Comment:
Maybe not necessary, but it might be good to explicitly say that these are for documentation only and do not do any validation. For example, `.format('uuid')` will not enforce that the given input is actually formatted like a UUID.

Reply:
Good call!

Reply:
Fixed!